_posts/2019-12-20-differential-privacy/differential_privacy.Rmd

```diff
@@ -0,0 +1,470 @@
+
+The values for `l2_norm_clip` and `noise_multiplier` chosen here follow those used in the [tutorials in the TF Privacy repo](https://github.com/tensorflow/privacy/tree/master/tutorials).
+
+Nicely, TF Privacy comes with a script that allows one to compute the attained $\epsilon$ beforehand, based on number of training examples, `batch_size`, `noise_multiplier` and number of training epochs.^[There is an additional parameter, $\delta$, that allows for bounding the risk that the privacy guarantee does not hold. The [recommendation](https://github.com/tensorflow/privacy/tree/master/tutorials) is to set this to at most the inverse of the number of training examples, which in out case would mean <= ~ `1e-04`; the default setting is `1e-06` so we should be fine here.]
+
+Calling that [script](https://github.com/tensorflow/privacy/blob/master/tensorflow_privacy/privacy/analysis/compute_dp_sgd_privacy.py), and assuming we train for 20 epochs here as well, 
+
+```{python}
+python compute_dp_sgd_privacy.py --N=3989 --batch_size=32 --noise_multiplier=1.1 --epochs=20
+```
+
+this is what we get back:
+
+```
+DP-SGD with sampling rate = 0.802% and noise_multiplier = 1.1 iterated over
+2494 steps satisfies differential privacy with eps = 2.73 and delta = 1e-06.
+```
+
+How good is a value of 2.73? Citing the [TF Privacy authors](https://github.com/tensorflow/privacy/tree/master/tutorials):
+
+> $\epsilon$ gives a ceiling on how much the probability of a particular output can increase by including (or removing) a single training example. We usually want it to be a small constant (less than 10, or, for more stringent privacy guarantees, less than 1). However, this is only an upper bound, and a large value of epsilon may still mean good practical privacy.
+
+
+Obviously, choice of $\epsilon$ is a (challenging) topic by itself, and not something we can elaborate on in a post dedicated to the technical aspects of DP with TensorFlow.
+
+How would $\epsilon$ change if we trained for 50 epochs instead? (This is actually what we'll do, seeing that training results on the test set tend to jump around quite a bit.)
+
+```{python}
+python compute_dp_sgd_privacy.py --N=3989 --batch_size=32 --noise_multiplier=1.1 --epochs=60
+```
+
+```
+DP-SGD with sampling rate = 0.802% and noise_multiplier = 1.1 iterated over
+6233 steps satisfies differential privacy with eps = 4.25 and delta = 1e-06.
+```
+Having talked about its parameters, now let's define the DP optimizer:
+
+```{r}
+l2_norm_clip <- 1
+noise_multiplier <- 1.1
+num_microbatches <- tf$cast(batch_size, tf$int32)
+learning_rate <- 0.01
+
+optimizer <- priv$DPAdamGaussianOptimizer(
+  l2_norm_clip = l2_norm_clip,
+  noise_multiplier = noise_multiplier,
+  num_microbatches = num_microbatches,
+  learning_rate = learning_rate
+)
+
+```
+
+One other change to make for DP. As gradients are clipped on a per-sample basis, the optimizer needs to work with per-sample losses as well: 
```

Comment:
```suggestion
There is one other change to make for DP. As gradients are clipped on a per-sample basis, the optimizer needs to work with per-sample losses as well: 
```

Reply:
That first sentence was a fragment, the other way to go would be to use the original wording and end it with a colon. But, given the following sentence ends with a colon, I think it's easier to turn that first line into a sentence of its own.

Reply:
Hm, we had the same situation in the last post... Is it too informal (I know I'm always on the verge ;-)) or is just totally uncommon in English?